atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionManager.java

```diff
@@ -415,6 +419,35 @@ <T, C extends PreCommitCondition, E extends Exception> T runTaskWithConditionRea
     @Timed
     long getUnreadableTimestamp();
 
+    /**
+     * Returns a timestamp that is before any leased timestamps returned by the consumer on {@link TimestampLeaseAwareTransaction#preCommit(TimestampLeaseName, int, PreCommitAction)}
+     * for a {@code timestampLeaseName} in open transactions.
+     * <p>
+     * This is similar to {@link TransactionManager#getImmutableTimestamp()} as it returns a timestamp before timestamps
+     * in open transactions, but for the immutable timestamp the timestamps considered are start timestamps for open
+     * transactions, while for leased timestamps the timestamps considered are leased timestamps from the corresponding
+     * {@code timestampLeaseName} in open transactions.
+     * <p>
+     * If no transactions with a {@code timestampLeaseName} lock are open, this method returns a new fresh timestamp
+     * (i.e. equivalent to {@link TimelockService#getFreshTimestamp()}).
+     * <p>
+     * Consumers should fetch the leased timestamp outside of transactions that potentially use it - if fetching the
+     * leased timestamp inside a transaction, it's possible for the transaction's start timestamp < leased timestamp,
+     * meaning the transaction cannot read all data up to leased timestamp.
+     *
+     * @param leaseName the name of the lease the timestamps are bound to
+     * @return the timestamp that is before any timestamp returned by the consumer of {@link TimestampLeaseAwareTransaction#preCommit(TimestampLeaseName, int, PreCommitAction)}
+     * for open transactions.
+     */
+    @RestrictedApi(
+            explanation = "This API is only meant to be used by AtlasDb proxies that want to make use of the"
+                    + " performance improvements by tracking leased timestamps of open transactions."
+                    + " Misuse of this feature can cause correctness issues.",
+            link = "https://github.com/palantir/atlasdb/pull/7305",
+            allowedOnPath = ".*/src/test/.*", // Unsafe behavior in tests is ok.
+            allowlistAnnotations = {ReviewedRestrictedApiUsage.class})
+    long getLeasedTimestamp(TimestampLeaseName leaseName);
```

Comment:
getMinLeasedTimestamp? for consistency with other places in the repo?

Reply:
Makes sense done